src/libos/src/signal/do_sigreturn.rs

```diff
@@ -278,29 +289,87 @@ fn handle_signals_by_user(
 
-        // Save the floating point registers
-        if curr_user_ctxt.extra_context_ptr != ptr::null_mut() {
-            // Signal from exception handling
-            debug_assert!(matches!(curr_user_ctxt.extra_context, ExtraContext::Xsave));
-            let fpregs_area =
-                unsafe { (&*(curr_user_ctxt.extra_context_ptr as *mut XsaveArea)) }.get_fpregs();
-            ucontext.fpregs.copy_from_slice(fpregs_area.as_slice());
-            // Clear the floating point registers, since we do not need to recover this when this syscall return
-            curr_user_ctxt.extra_context_ptr = ptr::null_mut();
+        if curr_user_ctxt.extra_context_ptr == ptr::null_mut() {
+            match curr_user_ctxt.extra_context {
+                ExtraContext::Fpregs => {
+                    // If curr_user_ctxt.extra_context_ptr is null, signal is handled in a normal syscall
+                    // We need a correct fxsave structure in the buffer, because the app may modify part of
+                    // it to update the floating point after the signal handler finished.
+                    // fpregs context is saved at two places:
+                    // (1) Copy to ucontext which can be used by user's handler
+                    // (2) Save at the heap and store the pointer to Occlum's defined CpuContext
+                    let fpregs = Box::new(FpRegs::save());
+                    ucontext.fpregs.copy_from_slice(fpregs.as_slice());
+                    ucontext.uc_mcontext.inner.extra_context_ptr = Box::into_raw(fpregs) as *mut u8;
+                    ucontext.uc_mcontext.inner.extra_context = ExtraContext::Fpregs;
+                    ucontext.uc_mcontext.inner.extra_context_size =
+                        std::mem::size_of::<FpRegs>() as u64;
+                }
+                ExtraContext::XsaveOnStack | ExtraContext::XsaveOnHeap => unreachable!(),
+            }
         } else {
-            // Raise the signal with direct syscall
-            debug_assert!(
-                matches!(curr_user_ctxt.extra_context, ExtraContext::Fpregs)
-                    && curr_user_ctxt.extra_context_ptr == ptr::null_mut()
-            );
-
-            // We need a correct fxsave structure in the buffer,
-            // because the app may modify part of it to update the
-            // floating point after the signal handler finished.
-            let fpregs = FpRegs::save();
-            ucontext.fpregs.copy_from_slice(fpregs.as_slice());
+            // curr_user_ctxt.extra_context_ptr != ptr::null_mut()
+            match curr_user_ctxt.extra_context {
+                ExtraContext::Fpregs => {
+                    // When sysret, use fpregs context from the latest signal handler. Thus free old fpregs to
+                    // prevent memory leakage.
+                    drop(unsafe { Box::from_raw(curr_user_ctxt.extra_context_ptr as *mut FpRegs) });
+                    let fpregs = Box::new(FpRegs::save());
```

Comment:
In what case do you need to release Fpregs first and then re-new?

Reply:
Reuse the box ptr instead of creating new one. Fixed.